base/strings/util.jl

```diff
@@ -388,13 +388,83 @@ _replace(io, repl, str, r, pattern) = print(io, repl)
 _replace(io, repl::Function, str, r, pattern) =
     print(io, repl(SubString(str, first(r), last(r))))
 
-replace(str::String, pat_repl::Pair{Char}; count::Integer=typemax(Int)) =
-    replace(str, equalto(first(pat_repl)) => last(pat_repl); count=count)
-replace(str::String, pat_repl::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}};
-        count::Integer=typemax(Int)) =
-    replace(str, occursin(first(pat_repl)) => last(pat_repl), count)
+# replace Char with function that compares with this Char
+pairmap(p::Pair{Char}) = equalto(p.first) => p.second
+
+# replace collection of Char with function that checks occurrence in this collection
+function pairmap(p::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}})
+    occursin(p.first) => p.second
+end
+pairmap(p::Pair) = p
+
+# find the first of a list of patterns.
+# if several match at the same start position prefer the longer one
+# if also the length is equal, prefer the first in list
+# return value of the successful search and the pair
+
+function findnextall(pairs::NTuple{N,Pair}, s::AbstractString, st::Integer=1) where N
+    ps = start(pairs)
+    while !done(pairs, ps)
+        p, ps = next(pairs, ps)
+        r = findnext(p.first, s, st)
+        fr = first(r)
+        if fr > 0
+            minstart, minend, minp = fr, last(r), p
+            while !done(pairs, ps)
+                p, ps = next(pairs, ps)
+                r = findnext(p.first, s, st)
+                i, k = first(r), last(r)
+                if i > 0 && ( i < minstart || i == minstart && k > minend)
+                    minstart, minend, minp = i, k, p
+                end
+            end
+            return minstart:minend, minp
+        end
+    end
+    0:-1, (""=>"")
+end
+
+# cover the multiple pairs case
+
+function replace(str::String, pat_repls::Pair...; count::Integer=typemax(Int))
+    if count == 0 || length(pat_repls) == 0
+        return str
+    end
+    count < 0 && throw(DomainError(count, "`count` must be non-negative."))
+    pat_repls = pairmap.(pat_repls)
```

Comment:
Is this type-stable?

Reply:
Yes, I rely on the broadcasting dot.
Looks so, for me. Input is a tuple of `Pair`, as output is:
```
julia> pp(p::Pair{Char}) = equalto(p.first)=>p.second
pp (generic function with 3 methods)

julia> pp(p::Pair) = p.first=>p.second
pp (generic function with 3 methods)

julia> pp.( ('a'=>'b', "xyz"=>"") )
(Base.EqualTo{Char}('a') => 'b', "xyz" => "")

julia> typeof(pp.( ('a'=>'b', "xyz"=>"") ))
Tuple{Pair{Base.EqualTo{Char},Char},Pair{String,String}}

julia> pp.( ('a'=>'b', "xyz"=>"") ) isa Tuple{Pair}
false

julia> pp.( ('a'=>'b', "xyz"=>"") ) isa NTuple{N,Pair} where N
true
```

Reply:
That should work, but run `@code_warntype` on the `replace` method to make sure that's the case.

Reply:
```
julia> using Unicode

julia> replace("abcde", 'a'=>'Ä', ['b', 'c']=>uppercase)
"ÄBCde"

julia> @code_warntype replace("abcde", 'a'=>'Ä', ['b', 'c']=>uppercase)
Variables:
  str::String
  pat_repls::Tuple{Pair{Char,Char},Pair{Array{Char,1},typeof(uppercase)}}

Body:
  begin
      SSAValue(6) = (Core.getfield)(pat_repls::Tuple{Pair{Char,Char},Pair{Array{Char,1},typeof(uppercase)}}, 1)::Pair{Char,Char}
      SSAValue(7) = (Core.getfield)(pat_repls::Tuple{Pair{Char,Char},Pair{Array{Char,1},typeof(uppercase)}}, 2)::Pair{Array{Char,1},typeof(uppercase)}
      SSAValue(2) = $(Expr(:invoke, MethodInstance for #replace#308(::Int64, ::Function, ::String, ::Pair{Char,Char}, ::Vararg{Pair,N} where N), :(Base.:(#replace#308)), 9223372036854775807, replace, :(str), SSAValue(6), SSAValue(7)))::String
      return SSAValue(2)
  end::String
```

Reply:
Ah, damn keyword arguments, they hide the body of the function. I guess the simplest approach is to remove the `count` argument for the check.

Reply:
I updated the output of `@code_warntype` after removing the keyword argument.

There are test cases missing, which use this argument :_)

Reply:
OK, looks good.

Reply:
Wait, the output looks incomplete.  Some variables do not appear to be inferred as concrete types, but they don't appear in the body.